core/arch/arm/kernel/boot.c

```diff
@@ -1133,7 +1133,7 @@ static void discover_nsec_memory(void)
 	const struct core_mmu_phys_mem *mem_begin = NULL;
 	const struct core_mmu_phys_mem *mem_end = NULL;
 	size_t nelems;
-	void *fdt = get_external_dt();
+	void *fdt = get_dt();
```

Comment:
This change would break platforms that embed a DTB in core but that expects REE memory areas are described from REE DTB (currently the external DT), not OP-TEE DTB.
I would have rather see a scheme like getting the secure DTB, find memory nodes and if none found fallback to external DTB to look for memory nodes.

Reply:
Thanks @etienne-lms, I'm still trying to get all the DT stuff into my head clearly as the docs are a big vague or I'm overlooking the details.  From the code I think we are saying that non-secure memory, right now, must always come from the external device tree and there is no option for it to come from the embedded DT and maybe that makes sense.  This implies that you would typically use both CFG_EMBED_DT and an external DT (CFG_DT_ADDR) at the same time, true?

Reply:
I also see some drivers pulling driver information from only the embedded DT and failing if it's not there.

Reply:
> From the code I think we are saying that non-secure memory, right now, must always come from the external device tree and there is no option for it to come from the embedded DT and maybe that makes sense.

Sorry I was not clear on that. IMO location of the non-secure memory normal world could use as TEE/REE shared memory could be retrieved from either the embedded/secure DT or the external/non-secure DT.
I think the implementation should allow OP-TEE to find this information from the secure DT and if not found then fallback to fetch it from the non-secure DT.
My question is how should the secure DT distinguish whether memory nodes define secure memory address ranges or overall system memory address ranges (whatever it is secure or not).

> This implies that you would typically use both CFG_EMBED_DT and an external DT (CFG_DT_ADDR) at the same time, true?

Yes, platforms could rely on both the embedded/secure DT and an external/non-secure DT. Secure config information are read from the secure DT. Non-secure information (mainly debug stuf) could be read from the non-secure DT.

This is my view, feel free to argue, there are rooms for changes I think.